File: Android~/plugin/src/main/java/com/mopsicus/umi/Bridge.java
package com.mopsicus.umi;


import android.util.Log;

import com.unity3d.player.UnityPlayer;

import org.json.JSONException;
import org.json.JSONObject;

public class Bridge {

    /**
     * GameObject on scene in Unity with handler
     */
    String object = "MobileInput";

    /**
     * Method name which will be process messages from plugin
     */
    String receiver = "OnDataReceive";

    /**
     * Flag to on/off debug log
     */
    boolean isDebug = false;

    /**
     * Send data in JSON format to Unity
     *
     * @param data   Data to send in Unity
     */
    @SuppressWarnings("unused")
    public void sendData(String data) {
        JSONObject info = new JSONObject();
        try {
            info.put("data", data);
        } catch (JSONException e) {
            if (isDebug) {
                Log.e("[UMI]", String.format("send data error: %s", e));
            }
        }
        UnityPlayer.UnitySendMessage(object, receiver, info.toString());
    }

    /**
     * Send error code without data
     *
     * @param code   Error code
     */
    @SuppressWarnings("unused")
    public void sendError(String code) {
        sendError(code, "");
    }

    /**
     * Send error in JSON format to Unity
     *
     * @param code   Error code
     * @param data   Error data
     */
    public void sendError(String code, String data) {
        JSONObject error = new JSONObject();
        JSONObject info = new JSONObject();
        try {
            error.put("code", code);
            error.put("message", data);
            info.put("error", error);
        } catch (JSONException e) {
            if (isDebug) {
                Log.e("[UMI]", String.format("send error error: %s", e));
            }
        }
        UnityPlayer.UnitySendMessage(object, receiver, info.toString());
    }

    /**
     * Init bridge
     *
     * @param object Object to send data
     * @param receiver Method name
     * @param isDebug On/off debug log
     */
    public void initialize(String object, String receiver, boolean isDebug) {
        this.object = object;
        this.receiver = receiver;
        this.isDebug = isDebug;
    }

}

File: Plugins/iOS/Bridge.m
#import "Bridge.h"

/// Bridge implementation to communicate with Unity app
@implementation Bridge

/// GameObject that will be receive messages
static NSString *object = @"MobileInput";

/// Method name in script
static NSString *receiver = @"OnDataReceive";

/// Flag to on/off debug log
static bool isDebug = NO;

/// Convert JSON string to NSDictionary
/// - Parameter json: JSON string
+ (NSDictionary *)jsonToDict:(NSString *)json {
    NSError *error;
    NSData *data = [json dataUsingEncoding:NSUTF8StringEncoding];
    return (error) ? NULL : [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingMutableContainers error:&error];
}

/// Convert NSDictionary to JSON string
/// - Parameter dict: NSDictionary to convert
+ (NSString *)dictToJson:(NSDictionary *)dict {
    NSError *error;
    NSData *data = [NSJSONSerialization dataWithJSONObject:dict options:0 error:&error];
    return (error) ? NULL : [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
}

/// Init plugin
/// - Parameter data: JSON string with receiver and method
+ (void)initialize:(NSString *)data {
    NSDictionary *params = [self jsonToDict:data];
    object = [params valueForKey:@"object"];
    receiver = [params valueForKey:@"receiver"];
    isDebug = [params valueForKey:@"debug"];
}

/// Send data from plugin to Unity app
/// - Parameter data: JSON string
+ (void)sendData:(NSString *)data {
    NSMutableDictionary *dict = [[NSMutableDictionary alloc] init];
    [dict setValue:data forKey:@"data"];
    NSString *result = [self dictToJson:dict];
    UnitySendMessage([object cStringUsingEncoding:NSUTF8StringEncoding], [receiver cStringUsingEncoding:NSUTF8StringEncoding], [result cStringUsingEncoding:NSUTF8StringEncoding]);
}

/// Send error to Unity app
/// - Parameters:
///   - code: Error code
///   - data: Error description
+ (void)sendError:(NSString *)code data:(NSString *)data  {
    NSDictionary *error = [NSDictionary dictionaryWithObjectsAndKeys:code, @"code", data, @"message", nil];
    NSMutableDictionary *dict = [[NSMutableDictionary alloc] init];
    [dict setValue:error forKey:@"error"];
    NSString *result = [self dictToJson:dict];
    UnitySendMessage([object cStringUsingEncoding:NSUTF8StringEncoding], [receiver cStringUsingEncoding:NSUTF8StringEncoding], [result cStringUsingEncoding:NSUTF8StringEncoding]);
}

@end


